### PR TITLE
Network: Work around OVN lb-add bug that allows multiple Load_Balancer records to be created with same name

### PR DESCRIPTION
**Pull request by tomponline**
The OVN `lb-add` command that adds entries to the `Load_Balancer` OVN northbound table has a bug that in certain situations allows multiple records with the same name to be created.

This then causes issues with the `lb-del` command that does not support the scenario of duplicate records of the same name.

This fix has multiple facets:

1. Changes the way LXD deletes existing load balancer entries by searching for them by name and then deleting them by UUID, this supports multiple records of the same name existing.
2. Changes the way LXD associates load balancer entries to a logical router by using the `add` database command directly rather than `lr-lb-add` that allows multiple records to exist with the same name.
3. Drops unused associated of load balancers to logical switches.

Fixes https://github.com/canonical/lxd/issues/13462

### CODE REVIEW

File: lxd/network/openvswitch/ovn.go [hunk 1]
```diff
@@ -1880,14 +1880,49 @@ func (o *OVN) PortGroupPortClearACLRules(portGroupName OVNPortGroup, portName OV
...
 func (o *OVN) LoadBalancerApply(loadBalancerName OVNLoadBalancer, routers []OVNRouter, switches []OVNSwitch, vips ...OVNLoadBalancerVIP) error {
 	lbTCPName := fmt.Sprintf("%s-tcp", loadBalancerName)
 	lbUDPName := fmt.Sprintf("%s-udp", loadBalancerName)
 
-	// Remove existing load balancers if they exist.
-	args := []string{"--if-exists", "lb-del", lbTCPName, "--", "lb-del", lbUDPName}
+	// Remove load balancers if they exist.
+	lbUUIDs, err := o.loadBalancerUUIDs(loadBalancerName)
+	if err != nil {
+		return fmt.Errorf("Failed getting UUIDs: %w", err)
```
Comment (roosterfish) [inline lxd/network/openvswitch/ovn.go:1914]: Would it be worth providing also the name of the LB in the error message? Something like 
`Failed getting UUIDs for load balancer %q: %w`?

File: lxd/network/openvswitch/ovn.go [hunk 2]
```diff
@@ -1987,15 +2024,19 @@ func (o *OVN) LoadBalancerDelete(loadBalancerNames ...OVNLoadBalancer) error {
 	var args []string
 
 	for _, loadBalancerName := range loadBalancerNames {
-		if len(args) > 0 {
-			args = append(args, "--")
+		lbUUIDs, err := o.loadBalancerUUIDs(loadBalancerName)
+		if err != nil {
+			return fmt.Errorf("Failed getting UUIDs: %w", err)
```
Comment (roosterfish) [inline lxd/network/openvswitch/ovn.go:2030]: Same as above regarding the comment on the LB's name.

File: lxd/network/openvswitch/ovn.go [hunk 3]
```diff
@@ -1946,6 +1946,10 @@ func (o *OVN) LoadBalancerApply(loadBalancerName OVNLoadBalancer, routers []OVNR
 			return fmt.Errorf("Missing VIP target(s)")
 		}
 
+		if len(args) > 0 {
+			args = append(args, "--")
```
Comment (roosterfish) [inline lxd/network/openvswitch/ovn.go:1947]: Could it be the case that the `--` gets added and a few lines down in `if r.Protocol == "udp"` the same gets appended again? 

I guess the resulting string would be `... -- -- lb-add ...`? Is this valid?

File: lxd/network/openvswitch/ovn.go [hunk 4]
```diff
@@ -1986,37 +1981,39 @@ func (o *OVN) LoadBalancerApply(loadBalancerName OVNLoadBalancer, routers []OVNR
...
+		var args []string
 
-			_, found = lbNames[lbUDPName]
-			if found {
-				args = append(args, "--", "lr-lb-add", string(r), lbUDPName)
-			}
+		// Get fresh list of load balancer UUIDs.
+		lbUUIDs, err := o.loadBalancerUUIDs(loadBalancerName)
+		if err != nil {
+			return fmt.Errorf("Failed getting UUIDs: %w", err)
```
Comment (roosterfish) [inline lxd/network/openvswitch/ovn.go:1999]: Again the comment about adding the LB's name.

**Review by roosterfish — CHANGES_REQUESTED** [review]
Hi, looks good, just a few smaller comments to clarify my understanding of the changes and to make the error messages more descriptive :)

File: lxd/network/openvswitch/ovn.go (hunk 1 above)
Comment (tomponline) [reply lxd/network/openvswitch/ovn.go:1914]: I was in two minds about this - the load balancer underlying concept is used for both LXD forwarders and load balancers, so I refrained on purpose from potentially exposing an error about the load balancers when it maybe in the context of a network forward. Instead I went for adding some context just about what phase of the apply process was failing.

Comment (roosterfish) [reply lxd/network/openvswitch/ovn.go:1914]: That makes total sense, fine for me then.

File: lxd/network/openvswitch/ovn.go (hunk 3 above)
Comment (roosterfish) [reply lxd/network/openvswitch/ovn.go:1947]: Maybe I am not fully aware of the command syntax here.

Comment (tomponline) [reply lxd/network/openvswitch/ovn.go:1947]: yeah i lose that bit

Comment (tomponline) [reply lxd/network/openvswitch/ovn.go:1947]: fixed

**Review by roosterfish — APPROVED** [review]
LGTM!

**Pull request merged**